### ISSUE TITLE
Flickering and laggy graphics on `master`

### ISSUE THREAD

**Issue by tpwrules** [CloudCompare/CloudCompare#1867]
**Describe the bug**
When compiling CloudCompare from source on [latest master](https://github.com/CloudCompare/CloudCompare/commit/e243efbb3e6ae65a96e42da7620fd46771f90fbe) on Linux, the window frequently flickers black as UI actions are performed. Additionally, dragging/rotating/zooming a large cloud is laggy, as if input data is being queued up when the graphics load gets heavy.

This effect is not present on the [2.13 Feb 27 alpha](https://github.com/CloudCompare/CloudCompare/commit/ef9d7deed9b3689f9c51b7b8df17e1b3474093a2) compiled from source on the same machine with the same libraries etc., nor 2.12.4.

**How to reproduce**
1. Open CloudCompare
2. Observe random black flickers
3. Observe that movement lags on large clouds

**Expected behaviour**
I expect these effects to be missing like earlier versions.

**Your environment**
 - CC Version (or git hash/tag):
 - OS & Version: Ubuntu 18.04 + Nixpkgs 23.05-unstable
 - Graphics card: NVidia Quadro P2000 + Driver 470.129.06
 - Qt Version (if compiling): Qt 5.15.10

**Comment by dgirardeau** [comment]
Hum, that's weird. Any chance that you compiled one of the versions in Debug mode?

Otherwise it may be due to the 'unified' version? https://github.com/CloudCompare/CloudCompare/commit/cce82ec26efdf3602689d6c751ed0353fae106c0

**Comment by tpwrules** [comment]
No, there's no chance of that. Both builds were identical except for the code checkout.

I checked and that is precisely the commit that introduces both issues. It's very easy to trigger the flickering by clicking a different cloud in the DB tree than the currently selected one. But sometimes mouse motion causes it too.

**Comment by dgirardeau** [comment]
I believe @rjanvier reproduced the issue. But finding the cure might be more complicated 😅

**Comment by tpwrules** [comment]
Thanks for letting me know. If I can help find the cause or cure I’d love to. CloudCompare is not usable with this problem!

**Comment by zjans** [comment]
Can confirm screen flickering (entire screen randomly every few seconds) after building CloudCompare from master (commit 36612f2ed1b43c75a53e1dce224e6f5e76fbaa73), there is no lag however and browsing/zooming is very responsive.

**my system**
OS & Version: Ubuntu 22.04.3 LTS on kernel 6.5.0-1008-oem
Graphics card: NVIDIA RTX A2000 12GB on 535.129.03
Qt Version (if compiling): Qt 5.15.3

**Comment by Loris-L** [comment]
@tpwrules @zjans Do you use xorg or wayland session? On my side (#1873) i have problems only in Xorg session, not in wayland one

**Comment by zjans** [comment]
I'm using Xorg, because wayland is not stable on my system with NVidia drivers.

**Comment by Loris-L** [comment]
True, up to Ubuntu 22.04 Nvidia and wayland don't get along. In next ubuntu lts tough (but already from ubuntu 23.04) nvidia drivers should work with wayland session, and it'll be better to use wayland

**Comment by Loris-L** [comment]
True, up to Ubuntu 22.04 Nvidia and wayland don't get along. In next ubuntu lts tough (but already from ubuntu 23.04) nvidia drivers should work with wayland session, and it'll be better to use wayland

**Comment by dgirardeau** [comment]
Thanks guys. So we'll focus mainly on the flickering issue for now.

I've made a clerical change yesterday (to avoid reimplementing a non-virtual function) but I don't believe it's related sadly: 0a171683b894e242aee254a981065dbf54171af5

**Comment by dgirardeau** [comment]
Ok, so I've installed Ubuntu 22.04 on a machine, compiled CC, and ran it.

On this machine, I don't have a real GPU (just an Intel Chipset).

I didn't see any flickering, but on the contrary, the Qt GUI update was 'blocked' as long as no refresh event of the 3D view was automatically triggered (for instance by the mouse pointer hovering it).

It led me to guess that the GL buffers were not swapped properly, or something was blocking the Qt OpenGL based composition mechanism... So I simply added a call to 'update()' at the end of the 'paintGL' routine. It fixed the issue on my setup.

However I am really unsure if it's related to your blinking issue (maybe to your 'lag' issue though, as you would have to wait a long time for this update to automatically trigger).

See the branch here: https://github.com/CloudCompare/CloudCompare/tree/linx_gl_update

**Comment by Loris-L** [comment]
@dgirardeau in my issue page i told you that yes, an update request at the end of paintGL solves the lag/block of mainwindow, but if i try to click on "open file" icon CC freezes and if I try to open a new 3D window and then interact with them there are still the lag/block problems. Could you try if with these action you have these same behaviours?
Could you also try if in wayland session there is any problem of these? On the bottom right of login screen session there is a gear that allows to switch between wayland and xorg.
Thanks very much for your support

**Comment by Loris-L** [comment]
> @dgirardeau in my issue page i told you that an update request at the end of doPaintGL in ccGLWindowInterface solves the lag/block of mainwindow, but if i try to click on "open file" icon CC freezes and if I try to open a new 3D window and then interact with them there are still the lag/block problems. Could you try if with these actions you have these same behaviours with your fix?

Can confirm these behaviours also with the fix in ccGLWindow

> Could you also try if in wayland session there is any problem of these without the update request in ccGLWindow? On the bottom right of login screen session there is a gear that allows to switch between wayland and xorg.

Also in wayland session there are lag problems without fix in ccGLWindow

**Comment by dgirardeau** [comment]
Ok, Romain found that the 3D view of ccViewer was working fine, and only CC was impacted. This led me to think that the issue was not coming from ccGLWindow(Interface) but more from how it was integrated with the QMdiArea widget.

And looking on the internet to similar issues, I found that the way the 'QSurfaceFormat' format is initialized can have an impact on how this work... And when I disabled the call to `setStereo(true)` (https://github.com/CloudCompare/CloudCompare/blob/master/libs/CCAppCommon/src/ccApplicationBase.cpp#L68), it fixed the issue on my side (without needing to 'update' anything).

Can you test the 'linx_gl_update' branch again? (I've reverted the previous commit, and replaced it by a new one).

I've also added a call to format.setSwapInterval(0); after reading a bug report (https://bugreports.qt.io/browse/QTBUG-46634) but I don't know what's its real effect is. So I would also be interested to know if it changes anything (in good or bad).

**Comment by Loris-L** [comment]
Very interesting, i'll test these changes and i'll update you

**Comment by Loris-L** [comment]
Hi @dgirardeau, i've only good news:

- I tried the changes in "linx_gl_update" branch both on my new laptop with Nvidia GPU (flickering problems) and on an old desktop PC with intel iGPU (mainwindow lag/block problems). It seems that there are no longer these problems, so the 2 types of problems are correlated.
- I also tried to compile CC without the call to `format.setSwapinterval(0)` and there aren't still lagging and flickering problems, so this big problem on linux OSes is only due to the call to `setStereo(true)`.
